Server.MirForms/Database/ItemInfoFormNew.cs

```diff
@@ -918,5 +927,55 @@ private void Gameshop_button_Click(object sender, EventArgs e)
 
             Envir.SaveDB();
         }
+
+        private void CellValueChanged(object sender, DataGridViewCellEventArgs e)
+        {
+            if (itemInfoGridView.CurrentCell is DataGridViewComboBoxCell ||
+                itemInfoGridView.CurrentCell is DataGridViewCheckBoxCell)
+            {
+                itemInfoGridView.Rows[e.RowIndex].Cells["Modified"].Value = true;
+            }
+        }
+
+        private void CurrentCellDirtyStateChanged(object sender, EventArgs e)
+        {
+            if (itemInfoGridView.IsCurrentCellDirty)
+            {
+                itemInfoGridView.CommitEdit(DataGridViewDataErrorContexts.Commit);
+            }
+        }
+
+        private void ItemInfoFormNew_FormClosing(object sender, FormClosingEventArgs e)
+        {
+            List<String> inError = new();
+            int indexColumn = itemInfoGridView.Columns["ItemIndex"].Index;
+            int nameColumn = itemInfoGridView.Columns["ItemName"].Index;
+
+            for (int i = 0; i < itemInfoGridView.RowCount; i++)
+            {
+                if (!String.IsNullOrEmpty(itemInfoGridView.Rows[i].ErrorText))
+                {
+                    inError.Add($"Index: [{itemInfoGridView.Rows[i].Cells[indexColumn].Value}] Item: [{itemInfoGridView.Rows[i].Cells[nameColumn].Value}]");
+                }
+            }
+
+            if (inError.Count > 0)
+            {
+                String msg = string.Join(Environment.NewLine, inError);
+                if (MessageBox.Show($"The following items are invalid: {msg}", "Discard Invalid Items?", MessageBoxButtons.OKCancel) == DialogResult.OK)
+                {
+                    SaveForm();
+                    Envir.SaveDB();
```

Comment:
i think you want a return here? Otherwise it'll save twice

Reply:
Change committed.